### ISSUE TITLE
CREATE's with failure condition

### ISSUE THREAD

**Issue by OlivierBBB** [Consensys/besu-sequencer-plugins#866]
> **Question.** If a `CREATE(2)` produces a failure condition does Besu create and enter a child frame ? In particular will it trigger `TraceContextEnter/TraceContextExit/TraceContextReEntry` ?

------

Our definition of **Failure condition**:
- no exception
- no abort (i.e. sufficient balance & creator call stack depth < 1024
- BUT: createe has either nonzero nonce and/or nonempty code

**Preliminary Conclusion.** The link you provide suggests that CREATE's that aren't aborted (and may therefore still raise a failure condition) **actually spawn a child context**.

**Note.** This would be easy to test by the way. Just attempt to CREATE2 with the same init code and salt two times in a row. The second CREATE2 will raise a failure condition.

------

@lu-pinto and @daniellehrner: could you confirm the above interpretation ?

**Comment by lu-pinto** [comment]
I don't think it gets to invoke the child frame because in the case of e.g. invalid init contract code, an empty item would be put on the [operand stack](https://github.com/hyperledger/besu/blob/main/evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCreateOperation.java#L118-L122) signaling failure and not a new frame. Another example would be if the [code is too large](https://github.com/hyperledger/besu/blob/main/evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCreateOperation.java#L106-L109) a halt state would be returned back. So there's some validation of the code before the child frame gets to execute from what I see, which would fall into your criteria I believe?

The child frame is only spawned [here](https://github.com/hyperledger/besu/blob/main/evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCreateOperation.java#L124) and then we re-enter the process cycle again to run the init code  and `TraceContextEnter` gets called.

**Comment by OlivierBBB** [comment]
Thanks @lu-pinto. WRT The code size issue

``` java
if (code != null && code.getSize() > maxInitcodeSize) { ... }
```
`maxInitcodeSize` limitations apply on later releases than London (IIRC it's twice as much as the maximum deployed bytecode size of 24576). Our arithmetization targets the London EVM.

**Comment by OlivierBBB** [comment]
> invalid init contract code

What do you mean by that ?

Also for info: everything in Linea is done with respect to the London EVM. So any EIP's beyond London don't apply.

**Comment by OlivierBBB** [comment]
This snippet of code you point to preceding the nonce update
```java
    if (value.compareTo(account.getBalance()) > 0
        || frame.getDepth() >= 1024
        || account.getNonce() == -1
        || code == null
        || code.getEofVersion() != frame.getCode().getEofVersion()) {
      fail(frame);
    }
```
 relates to what the issue calls **aborted CREATE's**.  In the present issue we are concerned with what we have called CREATE's raising a **failure condition**.

And the link you provide suggests that CREATE's that aren't aborted (and may therefore still raise a failure condition) _actually spawn a child context_.

**Comment by OlivierBBB** [comment]
**Sidenote (for myself)** I  recognize either
```java
        || account.getNonce() == -1
```
@letypequividelespoubelles pointed out that the nonce restriction is likely related to java's `Long.MAX_VALUE` and some EIP about the max value of nonces being `2**64 - 1`.

This condition is I guess some Besu stuff.
```java
        || code == null
```

**Note.** We are not enforcing the nonce to be a uint64 ... This may lead to some reference tests breaking.

**Comment by lu-pinto** [comment]
>Our arithmetization targets the London EVM.

I see... I don't remember hearing that until now. I'll have another look keeping that in mind.

> maxInitcodeSize limitations apply on later releases than London

Maybe I'm missing something but I don't see that this check is configurable through any settings specifically to the London hardfork. Currently the besu code that Linea is running on is built on a fairly recent version of the besu client. delivery-26 is built from backport of besu@529bd33 git hash. Maybe the condition doesn't trigger with a london genesis file but it's not explicitly apparent that is the case.

**Comment by lu-pinto** [comment]
I had another thorough look at unit tests that seem to confirm what I mentioned for `maxInitcodeSize` -> the check is done even for London EVMs. However the configuration is set to `Integer.MAX_VALUE` which is a large enough value that this check can be neglected. See london configuration [here](https://github.com/hyperledger/besu/blob/main/evm/src/main/java/org/hyperledger/besu/evm/EvmSpecVersion.java#L51).

Regarding the invalid init contract code case, when code is loaded before execution, you are right that only EVMs with EOF V1 spec are impacted (not london). There is only a tiny case when `maxEofVersion` ([CodeFactory.java](https://github.com/hyperledger/besu/blob/ec8429f800845f2af0728161119f77762f97f219/evm/src/main/java/org/hyperledger/besu/evm/code/CodeFactory.java#L69-L75)) but that is a developer error which shouldn't happen since it's defined in `EvmSpecVersion` also.

**Comment by lu-pinto** [comment]
Regarding:
>Preliminary Conclusion. The link you provide suggests that CREATE's that aren't aborted (and may therefore still raise a failure condition) actually spawn a child context.

I don't see anything else other than the contract account validation after loading the child frame on the stack, inside `ContractCreationProcessor.start()`. At this point we have already entered the context for the child frame, so your conclusion is correct.
The contract account is checked if it already exists [here](https://github.com/hyperledger/besu/blob/ec8429f800845f2af0728161119f77762f97f219/evm/src/main/java/org/hyperledger/besu/evm/processor/ContractCreationProcessor.java#L102). Is this the createe you were looking for @OlivierBBB? I'm guessing createe is the contract address being created by the sender. One other condition where we would fail is if the contract account storage is also not empty.
